docs/design/specs/runtime-async.md

```diff
@@ -59,31 +61,30 @@ Other restrictions are likely to be permanent, including
 * Suspension points may not appear in exception handling blocks.
 * Only four types will be supported as the return type for "runtime-async" methods: `System.Threading.Task`, `System.Threading.ValueTask`, `System.Threading.Task<T>`, and `System.Threading.ValueTask<T>`
 
-All async methods effectively have two entry points, or signatures. The first signature is the one present in method definitions: a `Task` or `ValueTask` returning method. The second signature is a "secondary signature", described in further detail in [I.8.6.1.5 Method signatures].
+In addition to the `AwaitAwaiter...` general-purpose methods defined above, the following specialty await methods are also defined.
+
+```C#
+  namespace System.Runtime.CompilerServices
+  {
+      public static class RuntimeHelpers
+      {
+          [MethodImpl(MethodImplOptions.Async)]
+          public static void Await(Task task);
+          [MethodImpl(MethodImplOptions.Async)]
+          public static void Await(ValueTask task);
+          [MethodImpl(MethodImplOptions.Async)]
+          public static T Await<T>(Task<T> task);
+          [MethodImpl(MethodImplOptions.Async)]
+          public static T Await<T>(ValueTask<T> task);
+      }
+  }
+```
 
-Callers may retrieve a Task/ValueTask return type from an async method via calling its primary, definitional signature. This functionality is available in both sync and async methods.
+These methods are also only legal to call inside async methods. These methods perform suspension like the `AwaitAwaiter...` methods, but are optimized for calling on the return value of a call to an async method. To achieve maximum performance, the IL sequence of two `call` instructions -- one to the async method and immediately one to the `Await` method -- should be preferred.
```

Comment:
How does this interact with ConfigureAwait? It seems like 99.999% of all use of await in our core libraries would not use these specialty methods?

Reply:
One of the things that the group proposed is that the runtime may be able to understand the pattern of `RuntimeHelpers.Await(taskMethod().ConfigureAwait(false))` and avoid the `Task` materialization in that case.

Reply:
Yes, but that doesn't quite fit with the proposed APIs.

Reply:
The main problem is that what comes out of `ConfigureAwait` isn't a task, it's a `ConfiguredTaskAwaitable`. The only way we have of dealing with this in the current design is to use the `AwaitAwaiter...` API.

I agree that configuration is something we should handle, it's just that it's not a 'new' problem with our API structure.

One option is to add an `Await` for `ConfigureTaskAwaitable` as above, and then recognize the triple sequence above.

Reply:
It seems to raise a question about what the value of this `Await` function is, over recognizing `RuntimeHelpers.AwaitAwaiterFromRuntimeAsync(Async2Function().GetAwaiter())` and `RuntimeHelpers.AwaitAwaiterFromRuntimeAsync(Async2Function().ConfigureAwait(constant).GetAwaiter())`.
Size comes to mind.

Reply:
I suppose `AwaitAwaiterFromRuntimeAsync` does not really have the right shape to be used in the way we'd like here, and it's not easy to generalize it to have a shape that would be usable either.

Reply:
Yes, good point. This was implicit, but let's make it explicit. I see the value of `await` as:
* Size of IL
* Complexity (a clearer sequence to recognize. If we open up GetAwaiter we need to also think about entire blocks like `if (!awaiter.IsCompleted) { awaiter.UnsafeAwaitAwaiterFromRuntimeAsync(); }`. `Await` is comparatively simpler
* Shorter/simpler sequence for the JIT

Reply:
That second point is particularly relevant. I expect that's the general pattern that Roslyn will emit for such locations.